flang/lib/Lower/Bridge.cpp

```diff
@@ -2508,8 +2531,30 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     }
   }
 
-  void genFIR(const Fortran::parser::CompilerDirective &) {
-    // TODO
+  void attachLoopDirective(const Fortran::parser::CompilerDirective &dir,
+                           Fortran::lower::pft::Evaluation *e) {
+    while (e->isDirective()) {
+      e = e->lexicalSuccessor;
+    }
+
+    if (e->isA<Fortran::parser::NonLabelDoStmt>()) {
+      e->dirs.push_back(&dir);
+    } else {
+      fir::emitFatalError(toLocation(),
+                          "loop directive must appear before a loop");
```

Comment:
I think this needs to be caught in semantics. Outputting here won't have the same format as other errors/warnings and won't be able to control them with the same machinery. Do other compilers produce warnings or errors when this directive is misplaced? It is not clear to me which would be correct.

Keep the fatal error here to catch bugs in semantics.

Reply:
Another point is how this combined with OpenMP/OpenACC loop directives. The vector directive could be placed before an OpenMP/openACC directive (or maybe after).

Since there is no real spec for this kind of scenarios, I think we may want to allow it and still reflect the directive intent on the ACC/OpenMP loop in FIR.

Reply:
I'm concerned about allowing this in the presence of OpenACC/OpenMP loops as I think that opens a whole can of worms for implementation. For what it's worth, clang doesn't allow the equivalent clang pragma in the presence of OpenMP directives:
```
❯ cat test.cpp
void foo(int* x, int n) {
#pragma clang loop vectorize(enable)
#pragma omp parallel for
        for (int i = 0; i < n; ++i)
                x[i] = i+i;

}
❯ clang++ -fopenmp test.cpp
test.cpp:3:1: error: expected a for, while, or do-while loop to follow '#pragma clang loop'
    3 | #pragma omp parallel for
      | ^
1 error generated.
```

Reply:
I have seen usage of loop related directives with OpenMP. If we are not fixing it here, can you add a TODO or "Not Yet Implemented message" for this case of mixing OpenMP and directives?

Reply:
Can you comment here?

Reply:
I have put a line in the Directives.md doc saying that this isn't implemented and should be added